lib/resty/mlcache.lua

```diff
@@ -384,26 +391,53 @@ local function set_lru(self, key, value, ttl, neg_ttl, l1_serializer)
 end
 
 
-local function set_shm(self, shm_key, value, ttl, neg_ttl)
+local function shm_set_retries(self, key, val, ttl, max_tries)
+    -- we will call `set()` N times to work around potential shm fragmentation.
+    -- when the shm is full, it will only evict about 30 to 90 items (via
+    -- LRU), which could lead to a situation where `set()` still does not
+    -- have enough memory to store the cached value, in which case we
+    -- try again to try to trigger more LRU evictions.
+
+    local tries = 0
+    local ok, err
+
+    while tries < max_tries do
+        tries = tries + 1
+
+        ok, err = self.dict:set(key, val, ttl)
+        if ok or err and err ~= "no memory" then
+            break
+        end
+    end
+
+    if not ok then
+        if err ~= "no memory" then
+            return nil, "could not write to lua_shared_dict '" .. self.shm
+                        .. "': " .. err
+        end
+
+        ngx_log(WARN, "could not write to lua_shared_dict '",
+                      self.shm, "' after ", tries, " tries (no memory), ",
+                      "it is either fragmented or cannot allocate more ",
+                      "memory, consider increasing 'opts.shm_set_tries'")
+    end
+
+    return true
+end
```

Comment:
Is it intentional to return `true` upon a `no memory` error? The original code had it also, but seems odd (also no mention in the docs about this)

Additionally you might want to return the number of tries executed, that would still have a truthy return value (non breaking), but also allow for some analytics reporting.

Reply:
> Is it intentional to return true upon a no memory error? The original code had it also, but seems odd (also no mention in the docs about this)

Yes, we don't want to unjustifiably hide the fetched value from the user's application.

> Additionally you might want to return the number of tries executed, that would still have a truthy return value (non breaking), but also allow for some analytics reporting.

Yeah we could, although tracking `forcible` would be more useful for this purpose. See #36